include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/exclusive_scan.hpp

```diff
@@ -0,0 +1,246 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/sycl.hpp>
+
+#include <oneapi/dpl/execution>
+#include <oneapi/dpl/numeric>
+
+#include <oneapi/dpl/async>
+
+#include <oneapi/dpl/internal/distributed_ranges_impl/concepts/concepts.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/detail/onedpl_direct_iterator.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/execution_policy.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/sp/allocators.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/sp/detail.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/sp/init.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/sp/vector.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/sp/views/views.hpp>
+
+namespace oneapi::dpl::experimental::dr::sp
+{
+
+template <typename ExecutionPolicy, distributed_contiguous_range R, distributed_contiguous_range O, typename U,
+          typename BinaryOp>
+void
+exclusive_scan_impl_(ExecutionPolicy&& policy, R&& r, O&& o, U init, BinaryOp binary_op)
+{
+    using T = stdrng::range_value_t<O>;
+
+    static_assert(std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, distributed_device_policy>);
+
+    auto zipped_view = views::zip(r, o);
+    auto zipped_segments = zipped_view.zipped_segments();
+
+    if constexpr (std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, distributed_device_policy>)
```

Comment:
Having both `static_assert` and `if constexpr` with the same condition is redundant; the first is enough, or even a constraint to the function. Or is it some work in progress that you plan to complete later?

The same applies to `inclusive_scan` and `reduce` implementations.

Reply:
#1730